src/mdl/holder.rs

```diff
@@ -0,0 +1,536 @@
+//! A module to handle presentation of an ISO 18013-5 mobile driving license (mDL).
+//!
+//! This module manages the session for presentation of an mDL, including generating
+//! the QR code and BLE parameters, as well as managing the types of information requested
+//! by the reader.
+//!
+//! The presentation process is begun by calling [`initialize_mdl_presentation`] and
+//! passing the id of the mdoc to be used as well as a UUID that the client
+//! will use for the BLE central client:
+//!
+
+use crate::common::*;
+use crate::{storage_manager::StorageManagerInterface, vdc_collection::VdcCollection};
+use std::ops::DerefMut;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use isomdl::{
+    definitions::{
+        device_engagement::{CentralClientMode, DeviceRetrievalMethods},
+        helpers::NonEmptyMap,
+        session, BleOptions, DeviceRetrievalMethod, SessionEstablishment,
+    },
+    presentation::device::{self, SessionManagerInit},
+};
+
+/// Begin the mDL presentation process for the holder when the desired
+/// Mdoc is already stored in a [VdcCollection].
+///
+/// Initializes the presentation session for an ISO 18013-5 mDL and stores
+/// the session state object in the device storage_manager.
+///
+/// Arguments:
+/// mdoc_id: unique identifier for the credential to present, to be looked up
+///          in the VDC collection
+/// uuid:    the Bluetooth Low Energy Client Central Mode UUID to be used
+///
+/// Returns:
+/// A Result, with the `Ok` containing a tuple consisting of an enum representing
+/// the state of the presentation, a String containing the QR code URI, and a
+/// String containing the BLE ident.
+///
+#[uniffi::export]
+pub async fn initialize_mdl_presentation(
+    mdoc_id: Uuid,
+    uuid: Uuid,
+    storage_manager: Arc<dyn StorageManagerInterface>,
+) -> Result<MdlPresentationSession, SessionError> {
+    let vdc_collection = VdcCollection::new(storage_manager);
+
+    let document = vdc_collection
+        .get(mdoc_id)
+        .await
+        .map_err(|_| SessionError::Generic {
+            value: "Error in VDC Collection".to_string(),
+        })?
+        .ok_or(SessionError::Generic {
+            value: "No credential with that ID in the VDC collection.".to_string(),
+        })?;
+
+    let mdoc = crate::credential::mdoc::Mdoc::from_cbor_encoded_document(
+        document.payload,
+        document.key_alias.ok_or(SessionError::Generic {
+            value: "key_alias missing in vdc collection".to_string(),
+        })?,
+    )
+    .map_err(|e| SessionError::Generic {
+        value: format!("Error retrieving MDoc from storage: {e:}"),
+    })?;
```

Comment:
This isn't wrong, but the intended method for transforming a VdcCollection credential into a known type is [`ParsedCredential::parse_from_credential`](https://github.com/spruceid/mobile-sdk-rs/blob/main/src/credential/mod.rs#L82-L92), or you can just do:
```suggestion
    let mdoc: Arc<Mdoc> = document.try_into()
    .map_err(|e| SessionError::Generic {
        value: format!("Error retrieving MDoc from storage: {e:}"),
    })?;
```

Reply:
Thanks, I've made that change!